contracts/StableJoeStaking.sol

```diff
@@ -0,0 +1,312 @@
+        require(isRewardToken[_rewardToken], "StableJoeStaking: token can't be removed");
+        updateReward(_rewardToken);
+        isRewardToken[_rewardToken] = false;
+        uint256 len = rewardTokens.length;
+        for (uint256 i; i < len; i++) {
+            if (rewardTokens[i] == _rewardToken) {
+                rewardTokens[i] = rewardTokens[len - 1];
+                rewardTokens.pop();
+                break;
+            }
+        }
+    }
+
+    /**
+     * @notice Set the deposit fee percent
+     * @param _depositFeePercent The new deposit fee percent
+     */
+    function setdepositFeePercent(uint256 _depositFeePercent) external onlyOwner {
+        require(_depositFeePercent < 5e17, "StableJoeStaking: deposit fee can't be greater than 50%");
+        depositFeePercent = _depositFeePercent;
+    }
+
+    /**
+     * @notice View function to see pending reward token on frontend
+     * @param _user The address of the user
+     * @return `_user`'s pending reward token
+     */
+    function pendingReward(address _user, IERC20Upgradeable _token) external view returns (uint256) {
+        require(isRewardToken[_token], "StableJoeStaking: wrong reward token");
+        UserInfo storage user = userInfo[_user];
+        uint256 totalJoe = internalJoeBalance;
+        uint256 _accRewardTokenPerShare = accRewardPerShare[_token];
+
+        uint256 rewardBalance;
+        if (_token == joe) rewardBalance = _token.balanceOf(address(this)).sub(internalJoeBalance);
+        else rewardBalance = _token.balanceOf(address(this));
+        if (rewardBalance != lastRewardBalance[_token] && totalJoe != 0) {
+            uint256 accruedReward = rewardBalance.sub(lastRewardBalance[_token]);
+            _accRewardTokenPerShare = _accRewardTokenPerShare.add(accruedReward.mul(PRECISION).div(totalJoe));
+        }
+        return user.amount.mul(_accRewardTokenPerShare).div(PRECISION).sub(user.rewardDebt[_token]);
+    }
+
+    /**
+     * @notice Withdraw JOE from sJOE and harvest the rewards
+     * @param _amount The amount of JOE to withdraw
+     */
+    function withdraw(uint256 _amount) external {
+        UserInfo storage user = userInfo[msg.sender];
+        uint256 previousAmount = user.amount;
+        require(previousAmount >= _amount, "StableJoeStaking: withdraw amount exceeds balance");
+        uint256 newAmount = user.amount.sub(_amount);
+        user.amount = newAmount;
+
+        uint256 len = rewardTokens.length;
+        uint256 pending;
+        if (previousAmount > 0) {
+            for (uint256 i; i < len; i++) {
+                IERC20Upgradeable token = rewardTokens[i];
+                updateReward(token);
+
+                pending = previousAmount.mul(accRewardPerShare[token]).div(PRECISION).sub(user.rewardDebt[token]);
+                if (pending != 0) safeTokenTransfer(token, msg.sender, pending);
+                user.rewardDebt[token] = newAmount.mul(accRewardPerShare[token]).div(PRECISION);
```

Comment:
Should we do this line before `safeTokenTransfer` to follow the checks/effects/interactions pattern.